File: apps/api/api/schemas.py
"""Pydantic schemas for API request/response and UI contracts.

Event packet models mirror docs/event_packet_spec.md as the contract of record.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


# --- Enums (match DB) ---
class UserRole(str, Enum):
    elder = "elder"
    caregiver = "caregiver"
    admin = "admin"


class SessionMode(str, Enum):
    offline = "offline"
    online = "online"


class RiskSignalStatus(str, Enum):
    open = "open"
    acknowledged = "acknowledged"
    dismissed = "dismissed"
    escalated = "escalated"


class FeedbackLabel(str, Enum):
    true_positive = "true_positive"
    false_positive = "false_positive"
    unsure = "unsure"


# --- Event packet (contract: docs/event_packet_spec.md) ---

# Supported payload schema versions; reject unsupported versions when strict.
SUPPORTED_PAYLOAD_VERSIONS: tuple[int, ...] = (1,)

EventTypeLiteral = Literal[
    "wake", "partial_asr", "final_asr", "intent",
    "tool_call", "tool_result", "tts", "error",
    "device_state", "watchlist_hit",
    "transaction_detected", "payee_added", "bank_alert_received",
]


class SpeakerRole(str, Enum):
    elder = "elder"
    agent = "agent"
    unknown = "unknown"


class SpeakerPayload(BaseModel):
    """Nested speaker: speaker_id?, role? (elder|agent|unknown)."""
    model_config = ConfigDict(extra="forbid")
    speaker_id: str | None = None
    role: SpeakerRole | None = None


class FinalAsrPayload(BaseModel):
    """final_asr payload: text?, text_hash?, lang?, confidence?, speaker?."""
    model_config = ConfigDict(extra="forbid")
    text: str | None = None
    text_hash: str | None = None
    lang: str | None = None
    confidence: float | None = Field(None, ge=0.0, le=1.0)
    speaker: SpeakerPayload | None = None


class IntentPayload(BaseModel):
    """intent payload: name required, slots?, confidence?."""
    model_config = ConfigDict(extra="forbid")
    name: str = Field(..., min_length=1)
    slots: dict[str, Any] = Field(default_factory=dict)
    confidence: float | None = Field(None, ge=0.0, le=1.0)


class DeviceStatePayload(BaseModel):
    """device_state payload: online required, battery?, wifi_ssid_hash?."""
    model_config = ConfigDict(extra="forbid")
    online: bool = ...
    battery: float | None = Field(None, ge=0.0, le=1.0)
    wifi_ssid_hash: str | None = None


class EventPacket(BaseModel):
    """
    Single event packet. Contract: docs/event_packet_spec.md.
    Strict validation; payload shape validated by event_type where defined.
    """
    model_config = ConfigDict(strict=True, extra="forbid")

    session_id: UUID
    device_id: UUID
    ts: datetime
    seq: int = Field(..., ge=0)
    event_type: EventTypeLiteral
    payload_version: int = Field(1, description="Version of payload schema")
    payload: dict[str, Any] = Field(default_factory=dict)

    @field_validator("payload_version")
    @classmethod
    def payload_version_supported(cls, v: int) -> int:
        if v not in SUPPORTED_PAYLOAD_VERSIONS:
            raise ValueError(f"payload_version {v} not in supported {SUPPORTED_PAYLOAD_VERSIONS}")
        return v

    @model_validator(mode="after")
    def validate_payload_by_event_type(self) -> "EventPacket":
        """Strict payload validation for known event types."""
        if not self.payload:
            return self
        et = self.event_type
        if et == "final_asr":
            FinalAsrPayload.model_validate(self.payload)
        elif et == "intent":
            IntentPayload.model_validate(self.payload)
        elif et == "device_state":
            DeviceStatePayload.model_validate(self.payload)
        # wake, partial_asr, tool_call, tool_result, tts, error, watchlist_hit,
        # transaction_detected, payee_added, bank_alert_received: allow dict
        return self


# Backward compatibility: ingest still accepts the same request shape.
IngestEventItem = EventPacket


class IngestEventsRequest(BaseModel):
    """Raw events list; each item validated to EventPacket in handler for rejection logging."""
    events: list[dict[str, Any]]


class IngestEventsResponse(BaseModel):
    ingested: int
    session_ids: list[UUID]
    last_ts: datetime | None
    rejected: int = 0
    rejection_reasons: list[str] = Field(default_factory=list)


# --- Household ---
class HouseholdMe(BaseModel):
    id: UUID
    name: str
    role: UserRole
    display_name: str | None


class ConsentDefaults(BaseModel):
    share_with_caregiver: bool | None = None
    share_text: bool | None = None
    allow_outbound_contact: bool | None = None
    escalation_threshold: int | None = Field(None, ge=1, le=5)


class CaregiverContactCreate(BaseModel):
    name: str
    relationship: str | None = None
    phone: str | None = None
    email: str | None = None
    quiet_hours: dict[str, Any] | None = None
    priority: int = 1


class OnboardRequest(BaseModel):
    """Body for POST /households/onboard: optional display name, household name, consent defaults."""
    display_name: str | None = None
    household_name: str | None = None
    consent_defaults: ConsentDefaults | None = None


# --- Sessions ---
class SessionListItem(BaseModel):
    id: UUID
    device_id: UUID
    started_at: datetime
    ended_at: datetime | None
    mode: SessionMode
    consent_state: dict[str, Any]
    summary_text: str | None = None


class SessionListResponse(BaseModel):
    sessions: list[SessionListItem]
    total: int


class EventListItem(BaseModel):
    id: UUID
    ts: datetime
    seq: int
    event_type: str
    payload: dict[str, Any]
    text_redacted: bool


class EventsListResponse(BaseModel):
    events: list[EventListItem]
    total: int
    next_offset: int | None


# --- Risk signals (UI contract) ---
class RiskSignalCard(BaseModel):
    """Compact card for list view."""
    id: UUID
    ts: datetime
    signal_type: str
    severity: int = Field(ge=1, le=5)
    score: float
    status: RiskSignalStatus
    summary: str | None = None
    model_available: bool | None = Field(None, description="True when GNN produced this signal; false or null when rule-only")


class SubgraphNode(BaseModel):
    id: str
    type: str
    label: str | None = None
    score: float | None = None


class SubgraphEdge(BaseModel):
    src: str
    dst: str
    type: str
    weight: float | None = None
    rank: int | None = None


class RiskSignalDetailSubgraph(BaseModel):
    nodes: list[SubgraphNode]
    edges: list[SubgraphEdge]


class RiskSignalDetail(BaseModel):
    """Full detail for UI: includes explanation subgraph."""
    id: UUID
    household_id: UUID
    ts: datetime
    signal_type: str
    severity: int
    score: float
    status: RiskSignalStatus
    explanation: dict[str, Any]
    recommended_action: dict[str, Any]
    subgraph: RiskSignalDetailSubgraph | None = None
    session_ids: list[UUID] = Field(default_factory=list)
    event_ids: list[UUID] = Field(default_factory=list)
    entity_ids: list[UUID] = Field(default_factory=list)


class RiskSignalListResponse(BaseModel):
    signals: list[RiskSignalCard]
    total: int


# --- Single risk scoring contract (pipeline, worker, agent) ---
class RiskScoreItem(BaseModel):
    """One risk score from the shared scoring service. model_available and embedding/model_subgraph presence are explicit."""
    node_type: str = "entity"
    node_index: int = Field(..., ge=0)
    score: float = Field(..., ge=0.0, le=1.0)
    signal_type: str | None = Field("relational_anomaly", description="e.g. relational_anomaly")
    embedding: list[float] | None = Field(None, description="Present only when model ran and returned embeddings")
    model_subgraph: dict[str, Any] | None = Field(None, description="Present only when model ran and explainer produced subgraph")
    model_available: bool = Field(False, description="True when this score came from the GNN; false for rule-only fallback")


class RiskScoringModelMeta(BaseModel):
    """Model provenance when GNN ran (for similar incidents, centroid watchlists)."""
    model_name: str | None = None
    checkpoint_id: str | None = None
    embedding_dim: int | None = None


class RiskScoringResponse(BaseModel):
    """Single contract for risk scoring used by pipeline, worker, and Financial Security Agent. No silent placeholders."""
    model_available: bool = Field(..., description="True when GNN ran and produced scores; false when model unavailable")
    scores: list[RiskScoreItem] = Field(default_factory=list)
    fallback_used: str | None = Field(None, description="e.g. 'rule_only' when caller added explicit rule-based scores after model_available=false")
    model_meta: RiskScoringModelMeta | None = Field(None, description="Set when model_available=True; model_name, checkpoint_id, embedding_dim")


class FeedbackSubmit(BaseModel):
    label: FeedbackLabel
    notes: str | None = None


# --- Watchlists ---
class EmbeddingCentroidProvenance(BaseModel):
    """Provenance for embedding_centroid watchlist (GNN-dependent)."""
    risk_signal_ids: list[str] = Field(default_factory=list)
    window_days: int | None = None


class EmbeddingCentroidPattern(BaseModel):
    """watch_type=embedding_centroid: L2-normalized centroid, cosine threshold. Hard dependency: no centroid when model didn't run."""
    metric: str = "cosine"
    threshold: float = Field(..., description="cosine similarity threshold (e.g. 0.82)")
    centroid: list[float] = Field(..., description="L2-normalized embedding vector")
    dim: int = Field(..., ge=1)
    model_name: str | None = None
    provenance: EmbeddingCentroidProvenance | None = None


class WatchlistItem(BaseModel):
    id: UUID
    watch_type: str
    pattern: dict[str, Any]
    reason: str | None
    priority: int
    expires_at: datetime | None
    model_available: bool | None = Field(None, description="True when watch_type=embedding_centroid (GNN-dependent); false/null for hash/keyword")


class WatchlistListResponse(BaseModel):
    watchlists: list[WatchlistItem]


# --- Device sync ---
class DeviceSyncRequest(BaseModel):
    device_id: UUID
    last_upload_ts: datetime | None = None
    last_upload_seq_by_session: dict[str, int] = Field(default_factory=dict)


class DeviceSyncResponse(BaseModel):
    watchlists_delta: list[WatchlistItem]
    last_upload_ts: datetime | None
    last_upload_seq_by_session: dict[str, int]
    last_watchlist_pull_at: datetime | None
    high_risk_mode: dict[str, Any] | None = Field(None, description="When playbook pushed device_high_risk_mode; device enacts")


# --- Weekly summary ---
class SimilarIncident(BaseModel):
    """One similar past incident for Similar Incidents panel (real embeddings only)."""
    risk_signal_id: UUID
    similarity: float = Field(description="Cosine similarity from real embedding")
    score: float = Field(description="Same as similarity (backward compat)")
    ts: datetime | None = None
    signal_type: str | None = None
    severity: int | None = None
    status: str | None = None
    label_outcome: str | None = Field(None, description="confirmed_scam | false_positive | open")
    outcome: str | None = Field(None, description="Same as label_outcome")


class RetrievalProvenance(BaseModel):
    """Provenance for similar-incidents retrieval (model_available path only)."""
    model_name: str | None = None
    checkpoint_id: str | None = None
    embedding_dim: int | None = None
    timestamp: datetime | None = None


class SimilarIncidentsResponse(BaseModel):
    """When model did not run, available=false and similar=[]; do not compute similarity on synthetic embeddings."""
    available: bool = Field(True, description="False when no embedding (e.g. model_not_run)")
    reason: str | None = Field(None, description="e.g. 'model_not_run' when embedding missing")
    similar: list[SimilarIncident] = Field(default_factory=list)
    retrieval_provenance: RetrievalProvenance | None = Field(None, description="Set when available=True; model_name, checkpoint_id, embedding_dim, timestamp")


class WeeklySummary(BaseModel):
    id: UUID
    period_start: datetime | None
    period_end: datetime | None
    summary_text: str
    summary_json: dict[str, Any]


# --- Outbound outreach (caregiver notify/call/email) ---
class OutreachActionRequest(BaseModel):
    risk_signal_id: UUID
    channel_preference: str | None = Field(None, description="sms | email | voice_call")
    dry_run: bool = Field(False, description="If true, preview only; no send, no DB write")


class OutboundActionPayloadElderSafe(BaseModel):
    """Elder-safe view: only elder_safe_message and status."""
    elder_safe_message: str | None = None
    status: str = "queued"
    created_at: datetime | None = None


class OutboundActionResponse(BaseModel):
    id: UUID
    household_id: UUID
    triggered_by_risk_signal_id: UUID | None
    action_type: str
    channel: str
    recipient_name: str | None
    recipient_contact: str | None = None  # Redacted for elder; caregiver sees full
    recipient_contact_last4: str | None = None  # Safe for display; never raw to elder
    payload: dict[str, Any]
    status: str  # queued | sent | delivered | failed | suppressed
    provider: str
    provider_message_id: str | None
    error: str | None
    created_at: datetime
    sent_at: datetime | None
    delivered_at: datetime | None


class OutreachSummaryCounts(BaseModel):
    sent: int = 0
    suppressed: int = 0
    failed: int = 0
    queued: int = 0
    delivered: int = 0


class OutreachSummaryRecentItem(BaseModel):
    id: UUID
    status: str
    created_at: datetime | None
    sent_at: datetime | None
    error: str | None
    triggered_by_risk_signal_id: UUID | None
    channel: str | None
    recipient_contact_last4: str | None


class OutreachSummaryResponse(BaseModel):
    counts: OutreachSummaryCounts
    recent: list[OutreachSummaryRecentItem]


# --- Capabilities (household) ---
class CapabilitiesMe(BaseModel):
    household_id: UUID
    notify_sms_enabled: bool
    notify_email_enabled: bool
    device_policy_push_enabled: bool
    bank_data_connector: str
    bank_control_capabilities: dict[str, Any]
    updated_at: datetime | None


class CapabilitiesPatch(BaseModel):
    notify_sms_enabled: bool | None = None
    notify_email_enabled: bool | None = None
    device_policy_push_enabled: bool | None = None
    bank_data_connector: str | None = None
    bank_control_capabilities: dict[str, Any] | None = None


# --- Action playbooks & tasks ---
class ActionTaskDetail(BaseModel):
    id: UUID
    playbook_id: UUID
    task_type: str
    status: str
    details: dict[str, Any]  # API redacts phone/email for elder
    completed_by_user_id: UUID | None
    completed_at: datetime | None
    created_at: datetime


class PlaybookDetail(BaseModel):
    id: UUID
    household_id: UUID
    risk_signal_id: UUID
    playbook_type: str
    graph: dict[str, Any]
    status: str
    created_at: datetime
    updated_at: datetime
    tasks: list[ActionTaskDetail] = Field(default_factory=list)


class PlaybookTaskCompleteRequest(BaseModel):
    notes: str | None = None


# --- Incident packets ---
class IncidentPacketResponse(BaseModel):
    id: UUID
    household_id: UUID
    risk_signal_id: UUID
    packet_json: dict[str, Any]
    created_at: datetime


# --- Incident response agent ---
class IncidentResponseRunRequest(BaseModel):
    risk_signal_id: UUID = Field(..., description="Risk signal to run incident response for")
    dry_run: bool = Field(False, description="If true, no DB write; return preview")

File: apps/web/src/lib/api/schemas.ts
/**
 * Zod schemas matching backend api_ui_contracts.md and apps/api/api/schemas.py
 */
import { z } from "zod";

const uuid = z.string().uuid();
const iso = z.string().datetime({ offset: true }).or(z.string());

export const UserRole = z.enum(["elder", "caregiver", "admin"]);
export type UserRole = z.infer<typeof UserRole>;

export const HouseholdMeSchema = z.object({
  id: uuid,
  name: z.string(),
  role: UserRole,
  display_name: z.string().nullable(),
});
export type HouseholdMe = z.infer<typeof HouseholdMeSchema>;

export const CapabilitiesMeSchema = z.object({
  household_id: uuid,
  notify_sms_enabled: z.boolean(),
  notify_email_enabled: z.boolean(),
  device_policy_push_enabled: z.boolean(),
  bank_data_connector: z.string(),
  bank_control_capabilities: z.record(z.string(), z.unknown()),
  updated_at: iso.nullable(),
});
export type CapabilitiesMe = z.infer<typeof CapabilitiesMeSchema>;

export const HouseholdConsentSchema = z.object({
  share_with_caregiver: z.boolean(),
  share_text: z.boolean(),
  allow_outbound_contact: z.boolean(),
  escalation_threshold: z.number(),
  updated_at: iso.nullable().optional(),
});
export type HouseholdConsent = z.infer<typeof HouseholdConsentSchema>;

export const SessionMode = z.enum(["offline", "online"]);
export const SessionListItemSchema = z.object({
  id: uuid,
  device_id: uuid,
  started_at: iso,
  ended_at: iso.nullable(),
  mode: SessionMode,
  consent_state: z.record(z.string(), z.unknown()),
  summary_text: z.string().nullable().optional(),
});
export type SessionListItem = z.infer<typeof SessionListItemSchema>;

export const SessionListResponseSchema = z.object({
  sessions: z.array(SessionListItemSchema),
  total: z.number(),
});
export type SessionListResponse = z.infer<typeof SessionListResponseSchema>;

export const EventListItemSchema = z.object({
  id: uuid,
  ts: iso,
  seq: z.number(),
  event_type: z.string(),
  payload: z.record(z.string(), z.unknown()),
  text_redacted: z.boolean(),
});
export type EventListItem = z.infer<typeof EventListItemSchema>;

export const EventsListResponseSchema = z.object({
  events: z.array(EventListItemSchema),
  total: z.number(),
  next_offset: z.number().nullable(),
});
export type EventsListResponse = z.infer<typeof EventsListResponseSchema>;

export const RiskSignalStatus = z.enum(["open", "acknowledged", "dismissed", "escalated"]);
export type RiskSignalStatus = z.infer<typeof RiskSignalStatus>;

export const RiskSignalCardSchema = z.object({
  id: uuid,
  ts: iso,
  signal_type: z.string(),
  severity: z.number().min(1).max(5),
  score: z.number(),
  status: RiskSignalStatus,
  summary: z.string().nullable().optional(),
  model_available: z.boolean().nullable().optional(),
});
export type RiskSignalCard = z.infer<typeof RiskSignalCardSchema>;

export const SubgraphNodeSchema = z.object({
  id: z.string(),
  type: z.string(),
  label: z.string().nullable().optional(),
  score: z.number().nullable().optional(),
});
export type SubgraphNode = z.infer<typeof SubgraphNodeSchema>;

export const SubgraphEdgeSchema = z.object({
  src: z.string(),
  dst: z.string(),
  type: z.string(),
  weight: z.number().nullable().optional(),
  rank: z.number().nullable().optional(),
});
export type SubgraphEdge = z.infer<typeof SubgraphEdgeSchema>;

export const RiskSignalDetailSubgraphSchema = z.object({
  nodes: z.array(SubgraphNodeSchema),
  edges: z.array(SubgraphEdgeSchema),
});
export type RiskSignalDetailSubgraph = z.infer<typeof RiskSignalDetailSubgraphSchema>;

export const RiskSignalDetailSchema = z.object({
  id: uuid,
  household_id: uuid,
  ts: iso,
  signal_type: z.string(),
  severity: z.number(),
  score: z.number(),
  status: RiskSignalStatus,
  explanation: z.record(z.string(), z.unknown()),
  recommended_action: z.record(z.string(), z.unknown()),
  subgraph: RiskSignalDetailSubgraphSchema.nullable().optional(),
  session_ids: z.array(uuid).optional().default([]),
  event_ids: z.array(uuid).optional().default([]),
  entity_ids: z.array(uuid).optional().default([]),
});
export type RiskSignalDetail = z.infer<typeof RiskSignalDetailSchema>;

export const RiskSignalListResponseSchema = z.object({
  signals: z.array(RiskSignalCardSchema),
  total: z.number(),
});
export type RiskSignalListResponse = z.infer<typeof RiskSignalListResponseSchema>;

export const FeedbackLabel = z.enum(["true_positive", "false_positive", "unsure"]);
export type FeedbackLabel = z.infer<typeof FeedbackLabel>;

export const SimilarIncidentSchema = z.object({
  risk_signal_id: uuid,
  similarity: z.number(),
  score: z.number(),
  ts: iso.nullable().optional(),
  signal_type: z.string().nullable().optional(),
  severity: z.number().nullable().optional(),
  status: z.string().nullable().optional(),
  label_outcome: z.string().nullable().optional(),
  outcome: z.string().nullable().optional(),
});
export type SimilarIncident = z.infer<typeof SimilarIncidentSchema>;

export const RetrievalProvenanceSchema = z.object({
  model_name: z.string().nullable().optional(),
  checkpoint_id: z.string().nullable().optional(),
  embedding_dim: z.number().nullable().optional(),
  timestamp: iso.nullable().optional(),
});
export type RetrievalProvenance = z.infer<typeof RetrievalProvenanceSchema>;

export const SimilarIncidentsResponseSchema = z.object({
  /** When false or omitted, similar incidents are not available (e.g. model not run). Do not default to true so backend value is respected. */
  available: z.boolean().optional(),
  reason: z.string().nullable().optional(),
  similar: z.array(SimilarIncidentSchema),
  retrieval_provenance: RetrievalProvenanceSchema.nullable().optional(),
});
export type SimilarIncidentsResponse = z.infer<typeof SimilarIncidentsResponseSchema>;

/** GET /risk_signals/{id}/page — compound alert detail (composable: detail + page + page_etag). Backend also returns flat keys for compat. */
export const RiskSignalPagePayloadSchema = z.object({
  detail: RiskSignalDetailSchema.optional(),
  page: z
    .object({
      session_events: z.array(EventListItemSchema).default([]),
      similar: SimilarIncidentsResponseSchema,
      actions: z.array(z.record(z.string(), z.unknown())).default([]),
      playbook: z
        .object({
          id: uuid,
          household_id: uuid,
          risk_signal_id: uuid,
          playbook_type: z.string(),
          graph: z.record(z.string(), z.unknown()),
          status: z.string(),
          created_at: iso,
          updated_at: iso,
          tasks: z.array(
            z.object({
              id: uuid,
              playbook_id: uuid,
              task_type: z.string(),
              status: z.string(),
              details: z.record(z.string(), z.unknown()),
              completed_by_user_id: uuid.nullable().optional(),
              completed_at: iso.nullable().optional(),
              created_at: iso,
            })
          ),
        })
        .nullable()
        .optional(),
      gating: z
        .object({
          investigation_refresh_allowed: z.boolean().default(false),
          investigation_refresh_reasons: z.array(z.string()).default([]),
          capabilities_snapshot: z.record(z.string(), z.unknown()).default({}),
        })
        .default({}),
    })
    .optional(),
  page_etag: z.string().nullable().optional(),
  risk_signal_detail: RiskSignalDetailSchema,
  similar_incidents: SimilarIncidentsResponseSchema,
  session_events: z.array(EventListItemSchema).default([]),
  outreach_actions: z.array(z.record(z.string(), z.unknown())).default([]),
  playbook: z
    .object({
      id: uuid,
      household_id: uuid,
      risk_signal_id: uuid,
      playbook_type: z.string(),
      graph: z.record(z.string(), z.unknown()),
      status: z.string(),
      created_at: iso,
      updated_at: iso,
      tasks: z.array(
        z.object({
          id: uuid,
          playbook_id: uuid,
          task_type: z.string(),
          status: z.string(),
          details: z.record(z.string(), z.unknown()),
          completed_by_user_id: uuid.nullable().optional(),
          completed_at: iso.nullable().optional(),
          created_at: iso,
        })
      ),
    })
    .nullable()
    .optional(),
  capabilities_snapshot: z.record(z.string(), z.unknown()).default({}),
  investigation_refresh_allowed: z.boolean().default(false),
  investigation_refresh_reasons: z.array(z.string()).default([]),
});
export type RiskSignalPagePayload = z.infer<typeof RiskSignalPagePayloadSchema>;

export const WatchlistItemSchema = z.object({
  id: uuid,
  watch_type: z.string(),
  pattern: z.record(z.string(), z.unknown()),
  reason: z.string().nullable(),
  priority: z.number(),
  expires_at: iso.nullable(),
  model_available: z.boolean().nullable().optional(),
});
export type WatchlistItem = z.infer<typeof WatchlistItemSchema>;

export const WatchlistListResponseSchema = z.object({
  watchlists: z.array(WatchlistItemSchema),
});
export type WatchlistListResponse = z.infer<typeof WatchlistListResponseSchema>;

export const WeeklySummarySchema = z.object({
  id: uuid,
  period_start: iso.nullable(),
  period_end: iso.nullable(),
  summary_text: z.string(),
  summary_json: z.record(z.string(), z.unknown()),
});
export type WeeklySummary = z.infer<typeof WeeklySummarySchema>;

/** WS message from server */
export const RiskSignalWsMessageSchema = z.object({
  type: z.literal("risk_signal"),
  id: z.string(),
  household_id: z.string(),
  ts: z.string(),
  signal_type: z.string(),
  severity: z.number(),
  score: z.number(),
});
export type RiskSignalWsMessage = z.infer<typeof RiskSignalWsMessageSchema>;

/** Outbound outreach (caregiver notify) */
export const OutreachActionSchema = z.object({
  id: uuid,
  household_id: uuid,
  triggered_by_risk_signal_id: uuid.nullable(),
  action_type: z.string(),
  channel: z.string(),
  recipient_name: z.string().nullable(),
  recipient_contact: z.string().nullable(),
  recipient_contact_last4: z.string().nullable(),
  payload: z.record(z.string(), z.unknown()),
  status: z.enum(["queued", "sent", "delivered", "failed", "suppressed"]),
  provider: z.string(),
  provider_message_id: z.string().nullable(),
  error: z.string().nullable(),
  created_at: iso.nullable(),
  sent_at: iso.nullable(),
  delivered_at: iso.nullable(),
});
export type OutreachAction = z.infer<typeof OutreachActionSchema>;

export const OutreachSummaryCountsSchema = z.object({
  sent: z.number(),
  suppressed: z.number(),
  failed: z.number(),
  queued: z.number(),
  delivered: z.number(),
});
export const OutreachSummaryRecentItemSchema = z.object({
  id: uuid,
  status: z.string(),
  created_at: iso.nullable(),
  sent_at: iso.nullable(),
  error: z.string().nullable(),
  triggered_by_risk_signal_id: uuid.nullable(),
  channel: z.string().nullable(),
  recipient_contact_last4: z.string().nullable(),
});
export const OutreachSummarySchema = z.object({
  counts: OutreachSummaryCountsSchema,
  recent: z.array(OutreachSummaryRecentItemSchema),
});
export type OutreachSummary = z.infer<typeof OutreachSummarySchema>;

// Protection (unified watchlists, rings, reports)
export const ProtectionWatchlistItemSchema = z.object({
  id: z.string(),
  category: z.string(),
  type: z.string(),
  display_label: z.string().nullable(),
  display_value: z.string().nullable(),
  explanation: z.string().nullable(),
  priority: z.number(),
  score: z.number().nullable(),
  source_agent: z.string().nullable(),
  evidence_signal_ids: z.array(z.string()).default([]),
});
export type ProtectionWatchlistItem = z.infer<typeof ProtectionWatchlistItemSchema>;

export const ProtectionWatchlistSummarySchema = z.object({
  total: z.number(),
  by_category: z.record(z.string(), z.number()),
  items: z.array(ProtectionWatchlistItemSchema).default([]),
});
export type ProtectionWatchlistSummary = z.infer<typeof ProtectionWatchlistSummarySchema>;

export const ProtectionRingSummarySchema = z.object({
  id: z.string(),
  household_id: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  score: z.number(),
  summary_label: z.string().nullable().optional(),
  summary_text: z.string().nullable().optional(),
  members_count: z.number().default(0),
  meta: z.record(z.string(), z.unknown()).default({}),
});
export type ProtectionRingSummary = z.infer<typeof ProtectionRingSummarySchema>;

export const ProtectionReportSummarySchema = z.object({
  kind: z.string(),
  last_run_at: z.string().nullable().optional(),
  last_run_id: z.string().nullable().optional(),
  summary: z.string().nullable().optional(),
  status: z.string().nullable().optional(),
});
export type ProtectionReportSummary = z.infer<typeof ProtectionReportSummarySchema>;

export const ProtectionOverviewSchema = z.object({
  watchlist_summary: ProtectionWatchlistSummarySchema,
  rings_summary: z.array(ProtectionRingSummarySchema).default([]),
  reports_summary: z.array(ProtectionReportSummarySchema).default([]),
  last_updated_at: z.string().nullable().optional(),
  data_freshness: z.record(z.string(), z.unknown()).default({}),
});
export type ProtectionOverview = z.infer<typeof ProtectionOverviewSchema>;

/** GET /protection/summary — counts + previews for dashboard cards */
export const ProtectionSummarySchema = z.object({
  updated_at: z.string().nullable().optional(),
  counts: z.object({ watchlists: z.number(), rings: z.number(), reports: z.number() }).default({ watchlists: 0, rings: 0, reports: 0 }),
  watchlists_preview: z.array(ProtectionWatchlistItemSchema).default([]),
  rings_preview: z.array(ProtectionRingSummarySchema).default([]),
  reports_preview: z.array(ProtectionReportSummarySchema).default([]),
});
export type ProtectionSummary = z.infer<typeof ProtectionSummarySchema>;

/** GET /protection/reports/latest — latest report metadata per type */
export const ProtectionReportsLatestSchema = z.object({
  updated_at: z.string().nullable().optional(),
  reports: z.record(
    z.string(),
    z.object({
      last_run_at: z.string().nullable().optional(),
      last_run_id: z.string().nullable().optional(),
      summary: z.string().nullable().optional(),
      status: z.string().nullable().optional(),
    })
  ),
});
export type ProtectionReportsLatest = z.infer<typeof ProtectionReportsLatestSchema>;
